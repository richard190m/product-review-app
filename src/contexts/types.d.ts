import { ReactNode } from 'react';

// Reviews
export interface ReviewProviderProps {
  children: ReactNode;
}

export interface Review {
  name: string;
  email: string;
  rating: number;
  title: string;
  comment: string;
  date: Date;
}

export interface HandleSubmit {
  // eslint-disable-next-line no-unused-vars
  (review: Review): void;
}

export interface ReviewsContextInterface {
  reviews: Review[];
  submit: HandleSubmit;
}

// Breakpoint
export interface BreakpointProviderProps {
  children: ReactNode;
}

export interface BreakpointContextInterface {
  breakpoint: string;
}
