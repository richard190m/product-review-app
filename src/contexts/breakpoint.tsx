import { BreakpointContextInterface, BreakpointProviderProps } from './types';
import { createContext, useEffect, useState } from 'react';

export const BreakpointContext =
  createContext<BreakpointContextInterface | null>(null);

export const BreakpointProvider = ({ children }: BreakpointProviderProps) => {
  const [breakpoint, setBreakpoint] = useState<string>('');

  const handleBreakpoint = () => {
    if (window.innerWidth < 768) breakpoint != 'xs' && setBreakpoint('xs');
    else if (window.innerWidth < 1024)
      breakpoint != 'sm' && setBreakpoint('sm');
    else if (window.innerWidth < 1440)
      breakpoint != 'md' && setBreakpoint('md');
    else breakpoint != 'lg' && setBreakpoint('lg');
  };

  useEffect(() => {
    handleBreakpoint();

    function handleResize() {
      handleBreakpoint();
    }

    window.addEventListener('resize', handleResize);

    return () => window.removeEventListener('resize', handleResize);
  }, []);

  return (
    <BreakpointContext.Provider value={{ breakpoint }}>
      {children}
    </BreakpointContext.Provider>
  );
};
