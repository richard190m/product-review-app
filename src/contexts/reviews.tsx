import { createContext, useEffect, useState } from 'react';
import {
  HandleSubmit,
  Review,
  ReviewProviderProps,
  ReviewsContextInterface,
} from './types';
import MockReviews from '../mocks/MockReviews';

export const ReviewsContext = createContext<ReviewsContextInterface | null>(
  null,
);

export const ReviewsProvider = ({ children }: ReviewProviderProps) => {
  const [reviews, setReviews] = useState<Review[]>([]);

  const handleSubmit: HandleSubmit = review => {
    setReviews([...reviews, review]);
  };

  useEffect(() => {
    // Load Reviews (mocks)
    setReviews(MockReviews);
  }, []);

  return (
    <ReviewsContext.Provider value={{ reviews, submit: handleSubmit }}>
      {children}
    </ReviewsContext.Provider>
  );
};
