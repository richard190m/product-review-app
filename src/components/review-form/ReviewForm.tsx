import {
  ButtonsContainer,
  ClearButton,
  Container,
  Error,
  FormStyled,
  InputStyled,
  LabelStyled,
  StarFilled,
  StarStyled,
  SubmitButton,
  Textarea,
  Title,
} from './styles';
import React, { useContext, useState } from 'react';
import PropTypes from 'prop-types';
import { Review } from '../../contexts/types';
import { ReviewsContext } from '../../contexts/reviews';

const INITIAL_STATE_REVIEW = {
  name: '',
  email: '',
  rating: 0,
  title: '',
  comment: '',
  date: new Date(),
};
const INITIAL_STATE_RATINGS = new Array(5).fill(false);

const ReviewForm = ({ visible = true }) => {
  const reviewsContext = useContext(ReviewsContext);
  const [ratings, setRatings] = useState(INITIAL_STATE_RATINGS);
  const [values, setValues] = useState<Review>(INITIAL_STATE_REVIEW);
  const [formError, setFormError] = useState('');

  const valideEmail = (email: string) => {
    const regExpEmail = new RegExp(/^[a-zA-Z0-9]+@[a-zA-Z0-9]+\.[A-Za-z]*$/);
    return regExpEmail.test(email);
  };

  const valideRanking = () => ratings.some(value => value === true);

  const isFormValid = () => {
    if (!valideEmail(values.email)) {
      setFormError('* Email is not valid!');
      return false;
    }

    if (!valideRanking()) {
      setFormError('* Ranking is required!');
      return false;
    }

    // set form error to initial state
    setFormError('');
    return true;
  };

  const cleanForm = () => {
    setValues(INITIAL_STATE_REVIEW);
    setRatings(INITIAL_STATE_RATINGS);
  };

  const handleFormSubmit = (event: React.FormEvent) => {
    event.preventDefault();

    if (isFormValid()) {
      const review = {
        ...values,
        rating: ratings.lastIndexOf(true) + 1,
        date: new Date(),
      };

      reviewsContext?.submit(review);

      // Clean form after submition
      cleanForm();
    }
  };

  const handleClickRating = (value: number) => {
    setRatings(ratings.map((rating, index) => index <= value));
  };

  const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setValues({ ...values, [event.target.name]: event.target.value });
  };

  const onChangeTextArea = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
    setValues({ ...values, [event.target.name]: event.target.value });
  };

  return (
    <Container show={visible} data-testid="review-form" id="review-form">
      <Title>Make a review</Title>
      <FormStyled onSubmit={handleFormSubmit} data-testid="form">
        <LabelStyled>
          Name:
          <InputStyled
            type="text"
            name="name"
            value={values.name}
            onChange={onChange}
            placeholder="Name"
            required
          />
        </LabelStyled>
        <LabelStyled>
          Email:
          <InputStyled
            type="email"
            name="email"
            value={values.email}
            onChange={onChange}
            placeholder="Email"
            required
          />
        </LabelStyled>
        <LabelStyled>
          Rating:
          <div>
            {ratings.map((value, index) =>
              value ? (
                <StarFilled
                  key={index}
                  data-testid={`ranking-${index}`}
                  onClick={() => handleClickRating(index)}
                />
              ) : (
                <StarStyled
                  key={index}
                  data-testid={`ranking-${index}`}
                  onClick={() => handleClickRating(index)}
                />
              ),
            )}
          </div>
        </LabelStyled>
        <LabelStyled>
          Title:
          <InputStyled
            type="text"
            name="title"
            value={values.title}
            onChange={onChange}
            placeholder="Comment Title"
            required
          />
        </LabelStyled>
        <LabelStyled>
          Comment:
          <Textarea
            name="comment"
            value={values.comment}
            onChange={onChangeTextArea}
            placeholder="Comment"
            rows={5}
            required
          />
        </LabelStyled>
        <Error data-testid="error-msg">{formError}</Error>
        <ButtonsContainer>
          <SubmitButton type="submit" value="Save review" />
          <ClearButton onClick={cleanForm}>Clear review</ClearButton>
        </ButtonsContainer>
      </FormStyled>
    </Container>
  );
};

ReviewForm.propTypes = {
  visible: PropTypes.bool,
};

export default ReviewForm;
