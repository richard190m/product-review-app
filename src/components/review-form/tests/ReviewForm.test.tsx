import { fireEvent, render } from '@testing-library/react';
import ReviewForm from '../ReviewForm';
import { ReviewsProvider } from '../../../contexts/reviews';

describe('ReviewsForm', () => {
  it('should render correctly', () => {
    const { container } = render(
      <ReviewsProvider>
        <ReviewForm />
      </ReviewsProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render error message when email or ranking not valid', () => {
    const { getByTestId, getByLabelText } = render(
      <ReviewsProvider>
        <ReviewForm />
      </ReviewsProvider>,
    );

    const form = getByTestId('form');

    fireEvent.change(getByLabelText('Email:'), {
      target: { value: 'test' },
    });
    fireEvent.submit(form);

    expect(getByTestId('error-msg')).toHaveTextContent('Email is not valid!');

    // Set valid email to trigger ranking error
    fireEvent.change(getByLabelText('Email:'), {
      target: { value: 'test@test.com' },
    });
    fireEvent.submit(form);

    expect(getByTestId('error-msg')).toHaveTextContent('Ranking is required!');

    // Set ranking to trigger error clean
    fireEvent.click(getByTestId('ranking-3'));
    fireEvent.submit(form);

    expect(getByTestId('error-msg')).toHaveTextContent('');
  });

  it('should submit form when all inputs with content', () => {
    const { getByTestId, getByLabelText } = render(
      <ReviewsProvider>
        <ReviewForm />
      </ReviewsProvider>,
    );

    const form = getByTestId('form');
    const inputName = getByLabelText('Name:');
    const inputEmail = getByLabelText('Email:');
    const inputTitle = getByLabelText('Title:');
    const inputComment = getByLabelText('Comment:');

    fireEvent.change(inputName, {
      target: { value: 'test' },
    });
    fireEvent.change(inputEmail, {
      target: { value: 'test@test.com' },
    });
    fireEvent.click(getByTestId('ranking-3'));
    fireEvent.change(inputTitle, {
      target: { value: 'review' },
    });
    fireEvent.change(inputComment, {
      target: { value: 'review comment' },
    });

    fireEvent.submit(form);

    // If valid, form submited and all inputs cleared
    expect(inputName).toHaveTextContent('');
    expect(inputEmail).toHaveTextContent('');
    expect(inputTitle).toHaveTextContent('');
    expect(inputComment).toHaveTextContent('');
  });
});
