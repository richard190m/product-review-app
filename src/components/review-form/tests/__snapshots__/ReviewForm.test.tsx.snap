// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewsForm should render correctly 1`] = `
<div>
  <div
    class="sc-bczRLJ iuLmvH"
    data-testid="review-form"
  >
    <form
      class="sc-hKMtZM cfGbOi"
      data-testid="form"
    >
      <label
        class="sc-ftvSup cTaIlw"
      >
        Name:
        <input
          class="sc-papXJ brxqLU"
          name="name"
          placeholder="Name"
          required=""
          type="text"
          value=""
        />
      </label>
      <label
        class="sc-ftvSup cTaIlw"
      >
        Email:
        <input
          class="sc-papXJ brxqLU"
          name="email"
          placeholder="Email"
          required=""
          type="email"
          value=""
        />
      </label>
      <label
        class="sc-ftvSup cTaIlw"
      >
        Rating:
        <div>
          <svg
            aria-hidden="true"
            class="StyledIconBase-ea9ulj-0 lbJwfL sc-dkzDqf flVVSf"
            data-testid="ranking-0"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21.947 9.179a1.001 1.001 0 0 0-.868-.676l-5.701-.453-2.467-5.461a.998.998 0 0 0-1.822-.001L8.622 8.05l-5.701.453a1 1 0 0 0-.619 1.713l4.213 4.107-1.49 6.452a1 1 0 0 0 1.53 1.057L12 18.202l5.445 3.63a1.001 1.001 0 0 0 1.517-1.106l-1.829-6.4 4.536-4.082c.297-.268.406-.686.278-1.065z"
            />
          </svg>
          <svg
            aria-hidden="true"
            class="StyledIconBase-ea9ulj-0 lbJwfL sc-dkzDqf flVVSf"
            data-testid="ranking-1"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21.947 9.179a1.001 1.001 0 0 0-.868-.676l-5.701-.453-2.467-5.461a.998.998 0 0 0-1.822-.001L8.622 8.05l-5.701.453a1 1 0 0 0-.619 1.713l4.213 4.107-1.49 6.452a1 1 0 0 0 1.53 1.057L12 18.202l5.445 3.63a1.001 1.001 0 0 0 1.517-1.106l-1.829-6.4 4.536-4.082c.297-.268.406-.686.278-1.065z"
            />
          </svg>
          <svg
            aria-hidden="true"
            class="StyledIconBase-ea9ulj-0 lbJwfL sc-dkzDqf flVVSf"
            data-testid="ranking-2"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21.947 9.179a1.001 1.001 0 0 0-.868-.676l-5.701-.453-2.467-5.461a.998.998 0 0 0-1.822-.001L8.622 8.05l-5.701.453a1 1 0 0 0-.619 1.713l4.213 4.107-1.49 6.452a1 1 0 0 0 1.53 1.057L12 18.202l5.445 3.63a1.001 1.001 0 0 0 1.517-1.106l-1.829-6.4 4.536-4.082c.297-.268.406-.686.278-1.065z"
            />
          </svg>
          <svg
            aria-hidden="true"
            class="StyledIconBase-ea9ulj-0 lbJwfL sc-dkzDqf flVVSf"
            data-testid="ranking-3"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21.947 9.179a1.001 1.001 0 0 0-.868-.676l-5.701-.453-2.467-5.461a.998.998 0 0 0-1.822-.001L8.622 8.05l-5.701.453a1 1 0 0 0-.619 1.713l4.213 4.107-1.49 6.452a1 1 0 0 0 1.53 1.057L12 18.202l5.445 3.63a1.001 1.001 0 0 0 1.517-1.106l-1.829-6.4 4.536-4.082c.297-.268.406-.686.278-1.065z"
            />
          </svg>
          <svg
            aria-hidden="true"
            class="StyledIconBase-ea9ulj-0 lbJwfL sc-dkzDqf flVVSf"
            data-testid="ranking-4"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21.947 9.179a1.001 1.001 0 0 0-.868-.676l-5.701-.453-2.467-5.461a.998.998 0 0 0-1.822-.001L8.622 8.05l-5.701.453a1 1 0 0 0-.619 1.713l4.213 4.107-1.49 6.452a1 1 0 0 0 1.53 1.057L12 18.202l5.445 3.63a1.001 1.001 0 0 0 1.517-1.106l-1.829-6.4 4.536-4.082c.297-.268.406-.686.278-1.065z"
            />
          </svg>
        </div>
      </label>
      <label
        class="sc-ftvSup cTaIlw"
      >
        Title:
        <input
          class="sc-papXJ brxqLU"
          name="title"
          placeholder="Comment Title"
          required=""
          type="text"
          value=""
        />
      </label>
      <label
        class="sc-ftvSup cTaIlw"
      >
        Comment:
        <textarea
          class="sc-jqUVSM kHqxWB"
          name="comment"
          placeholder="Comment"
          required=""
          rows="3"
        />
      </label>
      <p
        class="sc-eCYdqJ gubPie"
        data-testid="error-msg"
      />
      <div
        class="sc-jSMfEi cnrVoF"
      >
        <input
          class="sc-gKXOVf biPGXY"
          type="submit"
          value="Save review"
        />
        <button
          class="sc-iBkjds eqCWbf"
        >
          Clear review
        </button>
      </div>
    </form>
  </div>
</div>
`;
